IFC-1977: Add git config for user name and email, set config on repo

solababs opened the pull request:
This PR creates new git configurations that are then set on the repository. The created configurations are
- [x] `user_name` - The user name of the git user
- [x] `user_email` - The email of the git user

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * New Git configuration options: user name, user email, global config file path, and dev credential helper
  * Option to enable explicit non-fast-forward merge commits
  * Git global configuration is now initialized automatically during worker startup

* **Documentation**
  * Configuration reference updated with the new Git and development settings

* **Tests**
  * Added integration and unit tests validating Git user config and explicit-merge behavior

* **Changelog**
  * Entry documenting explicit merge commits and new Git settings

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

solababs committed:
IFC-1977: Add git config for user name and email, set config on repo

This PR creates new git configurations that are then set on the repository.
The created configurations are
- [x] `user_name` - The user name of the git user
- [x] `user_email` - The email of the git user

solababs committed:
update docs

solababs committed:
refactor test

ajtmccarty commented on backend/infrahub/git/base.py:
```diff
@@ -415,6 +417,13 @@ async def create_locally(
+        if config.SETTINGS.git.user_name and config.SETTINGS.git.user_email:
+            with repo.config_writer() as git_config:
+                git_config.set_value("user", "name", config.SETTINGS.git.user_name)
+                git_config.set_value("user", "email", config.SETTINGS.git.user_email)
```
is it possible to have a username without an email or vice versa?

ogenstad replied:
I think it would be possible if a someone just configures an override for one of them.

I'm not sure where this should be implemented. I was considering to add it as a global commit option when someone starts up the prefect worker. I.e. that they would just run "git config --global" once at that point during startup. Then we don't need to deal with it each time we do something locally in the repo. It would also allow users to change it globally with a config setting.

The annoying thing with this would be that if you ran the infrahub prefect worker locally on your laptop and had this option enabled it would change the global git config...

What do you guys think of this?

solababs replied:
That works, I can move the implementation if we are agreed on that

ajtmccarty replied:
sounds okay to me

dummyuser committed:
move setting git global config to async worker

solababs committed:
add removed import

solababs committed:
update test

solababs committed:
add return type

solababs committed:
fix mypy

solababs committed:
update test

solababs committed:
update test to mock git global config

solababs committed:
add custom git config file

solababs committed:
add log, use os for file creation

solababs committed:
refactor making directory

solababs committed:
Merge branch 'release-1.6' into sb-20251103-git-config-user-name-and-email-ifc-1977

ogenstad commented on backend/infrahub/config.py:
```diff
@@ -453,6 +454,8 @@ class GitSettings(BaseSettings):
         ),
     )
+    user_name: str | None = Field(default=None, description="User name of the git user")
+    user_email: EmailStr | None = Field(default=None, description="Email of the git user")
```
I think for the description we should clarify where it will be used as this is something that shows up in the documentation. I.e. we should state that the username and email provided here will be used when Infrahub commits code to a repository.

ogenstad commented on backend/infrahub/workers/infrahub_async.py:
```diff
@@ -40,6 +42,9 @@
 WORKER_DEFAULT_RESULT_STORAGE_BLOCK = f"redisstoragecontainer/{TASK_RESULT_STORAGE_NAME}"
 DEFAULT_TASK_LOGGERS = ["infrahub.tasks"]
+DEFAULT_GIT_GLOBAL_CONFIG_DIRECTORY = "/opt/infrahub"
+DEFAULT_GIT_GLOBAL_CONFIG_FILE = f"{DEFAULT_GIT_GLOBAL_CONFIG_DIRECTORY}/git-config"
```
I'd suggest that we also move this to the config object instead of having it as a hardcoded constant.

ogenstad commented on backend/infrahub/workers/infrahub_async.py:
```diff
@@ -204,3 +210,35 @@ async def _init_services(self, client: InfrahubClient) -> None:
+
+    async def set_git_global_config(self) -> None:
+        if INFRAHUB_PRODUCTION and not os.getenv("GIT_GLOBAL_CONFIG"):
+            os.makedirs(DEFAULT_GIT_GLOBAL_CONFIG_DIRECTORY, exist_ok=True)  # noqa: PTH103
```
Instead of ignoring the error here we should use Pathlib as suggested by the rule: https://docs.astral.sh/ruff/rules/os-makedirs/

ogenstad commented on backend/infrahub/workers/infrahub_async.py:
```diff
@@ -204,3 +210,35 @@ async def _init_services(self, client: InfrahubClient) -> None:
         self.service = service
+
+    async def set_git_global_config(self) -> None:
+        if INFRAHUB_PRODUCTION and not os.getenv("GIT_GLOBAL_CONFIG"):
```
As we discussed in Slack I think we should remove the `INFRAHUB_PRODUCTION` variable from here.

ogenstad commented on backend/tests/integration/workers/test_infrahubasync.py:
```diff
@@ -107,3 +112,64 @@ async def test_broken_flow(
+        user_email = "test@email.com"
+
+        await set_git_config("user.name", user_name)
+        await set_git_config("user.email", user_email)
```
As CodeRabbit indicates I'm wondering if this is correct, i.e. the idea would be that the test `test_worker_has_set_git_user_config()` would validate that the worker configures these values but here we have already added them in the fixture. Is this a leftover from a previous iteration?

ogenstad commented on backend/infrahub/workers/infrahub_async.py:
```diff
@@ -204,3 +210,35 @@ async def _init_services(self, client: InfrahubClient) -> None:
+                stdout=asyncio.subprocess.PIPE,
+                stderr=asyncio.subprocess.PIPE,
+            )
+            await proc_email.wait()
```
For these two commands we assume that everything will work fine. We should consider errors as well, what would it look like if any of these calls failed? Would the app crash or would the error be silently ignored?

solababs committed:
refactor tests

solababs committed:
Merge branch 'release-1.6' into sb-20251103-git-config-user-name-and-email-ifc-1977

solababs committed:
fix mypy

solababs committed:
update test typing

solababs committed:
refactor, update testsd

ogenstad commented on backend/infrahub/config.py:
```diff
@@ -453,6 +454,26 @@ class GitSettings(BaseSettings):
+        description=(
+            "Email of the git user. This will be used as the user email when Infrahub commits code to a repository"
+        ),
+    )
```
I wonder if we should set defaults here to match what we have in:
https://github.com/opsmill/infrahub/blob/stable/development/Dockerfile#L81-L82

That way we could validate the settings on each startup of the agent. Could also remove the need to set it within the Dockerfile.

ogenstad commented on backend/infrahub/workers/infrahub_async.py:
```diff
@@ -204,3 +207,48 @@ async def _init_services(self, client: InfrahubClient) -> None:
 
         self.service = service
+
+    async def set_git_global_config(self) -> None:
```
When we are creating a new config that will be used we'd also need to include the other settings, i.e. this config:
https://github.com/opsmill/infrahub/blob/stable/development/Dockerfile#L83-L85

ogenstad commented on backend/infrahub/config.py:
```diff
@@ -453,6 +454,26 @@ class GitSettings(BaseSettings):
+        ),
+    )
+    global_config_file: str | None = Field(
+        default="/opt/infrahub/.gitconfig",
```
As we set a default I don't thing it the setting should be an optional string. Using an environment variable I'm not sure we'd be able to override this and set it ot a `null` value.

solababs committed:
move setting git config from dockerfile to startup

ogenstad commented on backend/infrahub/workers/infrahub_async.py:
```diff
@@ -252,3 +252,52 @@ async def set_git_global_config(self) -> None:
+            "config",
+            "--global",
+            "credential.helper",
+            "/usr/local/bin/infrahub-git-credential",
```
I think this would actually be good to be able to control for local development, but I'm not sure if it's useful elsewhere. Could you add a git_credential_helper setting under the `DevelopmentSettings` config class and have the value default to `/usr/local/bin/infrahub-git-credential`?

fatih-acar commented on development/Dockerfile:
```diff
@@ -12,6 +12,8 @@ ENV PATH="${PATH}:/root/.local/bin" \
     INFRAHUB_DOCS_DIRECTORY="/opt/infrahub/docs"
 
+RUN groupadd -r -g 1000 infrahub && useradd --no-log-init -u 1000 --home-dir /opt/infrahub -r -g infrahub infrahub
+
```
I don't think the changes in `Dockerfile` are required.
Doing this is a breaking change for existing Infrahub instances regarding permissions (since the directories and files were created using `root`), I'd be careful with that... if you really need those changes.

ogenstad replied:
Ok, let's skip the changes to the Dockerfile.

solababs committed:
refactor running git config

solababs committed:
update docs

solababs committed:
revert docker change

solababs committed:
revert docker

solababs committed:
IFC-1979: Conditionally merge with no fast forward with `allow_explicit_merge_commit` git config (#7545)

* IFC-1977: Add git config for user name and email, set config on repo

This PR creates new git configurations that are then set on the repository.
The created configurations are
- [x] `user_name` - The user name of the git user
- [x] `user_email` - The email of the git user

* refactor test

* IFC-1979: Conditionally merge with no fast forward with `allow_explicit_merge_commit` git config

This PR adds a merge with no fast forward based on the setting from git.
Git setting is `allow_explicit_merge_commit`

* update docs

* use global git user config if repo config not set

* fix mypy

* move setting git global config to async worker

* remove repo config

* add removed import

* update setting name

* update test

* add return type

* fix mypy

* change commit message, change settting name

* remove unused function, update merge comment

---------

Co-authored-by: Test User <test@email.com>

solababs committed:
Merge branch 'release-1.6' into sb-20251103-git-config-user-name-and-email-ifc-1977

solababs committed:
add news fragment

solababs committed:
Merge branch 'sb-20251103-git-config-user-name-and-email-ifc-1977' of https://github.com/opsmill/infrahub into sb-20251103-git-config-user-name-and-email-ifc-1977

solababs committed:
fix markdown lint

solababs committed:
fix markdown lint

solababs committed:
fix markdown lint
